src/xrParticles/particle_actions_collection.cpp

```diff
@@ -1740,21 +1660,75 @@ void PATurbulence::Execute(ParticleEffect* effect, const float dt, float& tm_max
 
     age += dt;
 
-    u32 p_cnt = effect->p_count;
+    u32 count = effect->p_count;
 
-    if (!p_cnt)
+    if (!count)
         return;
 
-    TES_PARAMS tesParams;
-    tesParams.p_count = p_cnt; 
-    tesParams.effect = effect;
-    tesParams.offset = offset;
-    tesParams.age = age;
-    tesParams.epsilon = epsilon;
-    tesParams.frequency = frequency;
-    tesParams.octaves = octaves;
-    tesParams.magnitude = magnitude;
-    PATurbulenceExecuteStream(&tesParams);
+    auto processParticle = [&](Particle& m)
+    {
+        pVector pV;
+        pVector vX;
+        pVector vY;
+        pVector vZ;
+
+        pV.mad(m.pos, offset, age);
+        vX.set(pV.x + epsilon, pV.y, pV.z);
+        vY.set(pV.x, pV.y + epsilon, pV.z);
+        vZ.set(pV.x, pV.y, pV.z + epsilon);
+
+        float d = fractalsum3(pV, frequency, octaves);
+
+        pVector D;
+
+        D.x = fractalsum3(vX, frequency, octaves);
+        D.y = fractalsum3(vY, frequency, octaves);
+        D.z = fractalsum3(vZ, frequency, octaves);
+
+        __m128 _D = _mm_load_fvector(D);
+        __m128 _d = _mm_set1_ps(d);
+        __m128 _magnitude = _mm_set1_ps(magnitude);
+        __m128 _mvel = _mm_load_fvector(m.vel);
+        _D = _mm_sub_ps(_D, _d);
+        _D = _mm_mul_ps(_D, _magnitude);
+
+        __m128 _vmo = _mm_mul_ps(_mvel, _mvel); // _vmo = 00 | zz | yy | xx
+        __m128 _tmp = _mm_movehl_ps(_vmo, _vmo); // _tmp = 00 | zz | 00 | zz
+        _vmo = _mm_add_ss(_vmo, _tmp); // _vmo = 00 | zz | yy | xx + zz
+        _tmp = _mm_unpacklo_ps(_vmo, _vmo); // _tmp = yy | yy | xx + zz | xx + zz
+        _tmp = _mm_movehl_ps(_tmp, _tmp); // _tmp = yy | yy | yy | yy
+        _vmo = _mm_add_ss(_vmo, _tmp); // _vmo = 00 | zz | yy | xx + yy + zz
+        _vmo = _mm_sqrt_ss(_vmo); // _vmo = 00 | zz | yy | vmo
+
+        _mvel = _mm_add_ps(_mvel, _D);
+
+        __m128 _vmn = _mm_mul_ps(_mvel, _mvel); // _vmn = 00 | zz | yy | xx
+        _tmp = _mm_movehl_ps(_vmn, _vmn); // _tmp = 00 | zz | 00 | zz
+        _vmn = _mm_add_ss(_vmn, _tmp); // _vmn = 00 | zz | yy | xx + zz
+        _tmp = _mm_unpacklo_ps(_vmn, _vmn); // _tmp = yy | yy | xx + zz | xx + zz
+        _tmp = _mm_movehl_ps(_tmp, _tmp); // _tmp = yy | yy | yy | yy
+        _vmn = _mm_add_ss(_vmn, _tmp); // _vmn = 00 | zz | yy | xx + yy + zz
+        _vmn = _mm_sqrt_ss(_vmn); // _vmn = 00 | zz | yy | vmn
+
+        _vmo = _mm_div_ss(_vmo, _vmn); // _vmo = 00 | zz | yy | scale
+
+        _vmo = _mm_shuffle_ps(_vmo, _vmo, _MM_SHUFFLE(0, 0, 0, 0)); // _vmo = scale | scale | scale | scale
+        _mvel = _mm_mul_ps(_mvel, _vmo);
+
+        _mm_store_fvector(m.vel, _mvel);
+    };
+
+    constexpr u32 max_count = 75;
+
+    if (count < max_count)
```

Comment:
> Strange, but this cycle is always faster in multithreaded mode.

But this cycle is faster in singlethreaded mode for heavy particles...

Why?

Reply:
I think it could depend from the hardware features.

Reply:
What hardware and features do you mean?

Reply:
Unfortunately, I can't tell you the exact reason. Modern CPUs are too complex to fully understand the behavior of the code. The simplest example relates to caches in the processor. When some block of the code is fully located in the cache, the program works very fast, but when you tries to optimize the program (for example via loop unrolling) you could see a terrible performance drop (instead of expected performance improvements).